dao/src/main/java/se/fortnox/reactivewizard/db/transactions/Transaction.java

```diff
@@ -2,68 +2,41 @@
-import se.fortnox.reactivewizard.db.statement.Statement;
 
 import java.sql.Connection;
 import java.sql.SQLException;
-import java.util.Collection;
-import java.util.HashSet;
 import java.util.LinkedList;
-import java.util.Set;
+import java.util.List;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicReference;
 
 public class Transaction<T> {
 
     private static final Logger                                      log                 = LoggerFactory.getLogger(Transaction.class);
-    private final        AtomicReference<ConnectionProvider>         connectionProvider  = new AtomicReference<>();
-    private final        Iterable<Observable<T>>                     daoCalls;
-    private final        ConcurrentLinkedQueue<TransactionStatement> statementsToExecute = new ConcurrentLinkedQueue<>();
-    private final        Set<TransactionStatement>                   statementsToSubscribe = new HashSet<>();
+    private final        ConnectionProvider                          connectionProvider;
+    private final        ConcurrentLinkedQueue<TransactionStatement> statementsToExecute;
     private final        AtomicBoolean                               waitingForExecution = new AtomicBoolean(true);
 
-    Transaction(Iterable<Observable<T>> daoCalls) {
-        this.daoCalls = daoCalls;
+    Transaction(ConnectionProvider connectionProvider, List<TransactionStatement> statements) {
+        this.connectionProvider = connectionProvider;
+        this.statementsToExecute = new ConcurrentLinkedQueue<>(statements);
     }
 
-    void add(DaoObservable daoObservable) {
-        TransactionStatement transactionStatement = new TransactionStatement(this);
-        daoObservable.setTransactionStatement(transactionStatement);
-        statementsToExecute.add(transactionStatement);
-        statementsToSubscribe.add(transactionStatement);
-    }
-
-
-    public void execute() {
-        if (!isAllSubscribed()) {
-            // all DaoObservables have not been subscribed yet
-            throw new RuntimeException("Transaction execute called before all Observables were subscribed. This should never happen.");
-        }
-
-        if (isModifiedAfterCreation()) {
-            throw new RuntimeException("Transaction cannot be modified after creation.");
-        }
-
+    public void execute() throws Exception {
         if (!waitingForExecution.compareAndSet(true, false)) {
             return;
         }
 
-        if (connectionProvider.get() == null) {
-            throw new RuntimeException("No Connection Provider found!");
-        }
-
-        Connection connection = connectionProvider.get().get();
+        Connection connection = connectionProvider.get();
         try {
             executeTransaction(connection);
             closeConnection(connection);
-            allCompleted();
-        } catch (Throwable e) {
+        } catch (Exception e) {
```

Comment:
I think this needs to be Throwable to avoid that a non-exception error fails to rollback the transaction?